### PR TITLE
Bug fixes

### PR DESCRIPTION
**Pull request by apaszke**
* Fixed a number of clang warnings
* Fixed checks that picked ConvNd algos based on params (#1565)
* Indexing Variables with LongTensors now handles duplicate indices correctly (#1528)
* `Variable.type_as` now supports both tensor and variable arguments (#1163)

### CODE REVIEW

File: torch/csrc/autograd/variable.h [hunk 1]
```diff
@@ -22,7 +22,7 @@ struct Variable : std::enable_shared_from_this<Variable> {
     SavedVariable(const Variable& variable, Function* saved_for)
       : data(variable.data->clone_shallow())
       , grad_accumulator(variable.grad_accumulator)
-      , version(std::move(variable.version_counter->new_saved_ref()))
+      , version(variable.version_counter->new_saved_ref())
```
Comment (albanD) [inline torch/csrc/autograd/variable.h]: Why isn't the `std::move` needed here? `new_saved_ref` returns a `std::unique_ptr` no?

Comment (apaszke) [inline torch/csrc/autograd/variable.h]: Yes, but it's an rvalue and using `move` prevents copy elision (without it the compiler will just initialise this `version` in-place).

Comment (albanD) [inline torch/csrc/autograd/variable.h]: Ok

File: torch/autograd/variable.py [hunk 2]
```diff
@@ -63,8 +63,14 @@ def __getattr__(self, name):
...
     def __getitem__(self, key):
-        if isinstance(key, Variable) and type(key.data).__name__ == 'ByteTensor':
-            return MaskedSelect.apply(self, key)
+        if torch.is_tensor(key):
+            key = Variable(key)  # auto-wrap tensors
+        if isinstance(key, Variable):
+            if type(key.data).__name__ == 'ByteTensor':
+                return MaskedSelect.apply(self, key)
+            elif type(key.data).__name__ == 'LongTensor':
+                return IndexSelect.apply(self, 0, key)
```
Comment (ezyang) [inline torch/autograd/variable.py]: So, the reason that this is specific to Byte and Long is that these are the only "indexing" tensor types?

**Review by ezyang — APPROVED** [review]
LGTM

File: torch/autograd/variable.py (hunk 2 above)
Comment (apaszke) [inline torch/autograd/variable.py]: yes

**Review by colesbury — APPROVED** [review]
LGTM

**Pull request merged**